### DISCUSSION TITLE
Fedora

### DISCUSSION THREAD

**Discussion by aclater** [automatic-ripping-machine/automatic-ripping-machine#557] [General]
I recently got the latest version of ARM working on Fedora Server 36. I plan to do a clean install and document everything that needs to be done - mostly some little permission issues, but I'm happy to document.

Is anyone else out there running on a Red Hat based OS - RHEL, CentOS or Fedora?

**Comment by aclater** [comment]
I've begin working on an installation script for F36 - https://github.com/aclater/misc/blob/master/fedora-36-install.sh 

Naturally I built it from the main branch and not the dev - not sure how you guys feel about accepting that.

**Reply by iiPythonx** [reply]
Thank you for this! I was having issues getting ARM running within docker on Fedora, this worked perfectly first time.

**Reply by HaroldRobi** [reply]
I have ARM running under ubuntu server, so if you are having issues, follow the docker instructions provided here:  [https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/docker](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/docker)

**Comment by shitwolfymakes** [comment]
Anyone wanting to make ARM installable on a new distro and willing to handle it themselves is great in my book!

We are working up to a new release that will finally merge `v2_devel`'s more formalized file tree into `v2_master`, don't worry about PRing anything until after that's done.

For the moment, the most mature installation script is `ubuntu-20.04-install.sh` in the `v2_devel` branch. So using it's logical structure as a template will save you a lot of time figuring out what goes where. Our current manual instructions are woefully out of date and I will be updating them after the next release.

**Reply by aclater** [reply]
Cool. I'll take a look at that and see about making it fedora friendly.

**Reply by bentterp** [reply]
Hi! Did you succeed getting this to run on Fedora? I would be very interested

**Reply by PriamX** [reply]
I've been running ARM on Fedora since core 33. I'm currently running it as a native installation (no container, straight OS install) on core 37, and I'm also running it in a docker container on a Fedora core 35 system (planning to upgrade the OS soon).

For the native installation on Fedora I follow the [Ubuntu Instructions](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Setting-up-ARM-manually-(Ubuntu)) for the most part it works quite well. Of course, instead of `apt` on Ubuntu we use `dnf` on Fedora. And the package names may not be the same but a `dnf search <keyword>` helps find the right one for Fedora. 

I made some really rough notes at the time I installed it on core 33, I can share those with you via PM, they won't be entirely applicable anymore but perhaps we can work through them together, and share the end result with the community.

Edit: Fixed the Ubuntu Instructions link.

**Reply by PriamX** [reply]
I just had a peek at `scripts/ubuntu-20.04-install.sh` last updated 2 weeks ago in the main branch, and it looks pretty good. Perhaps even better than the Ubuntu instructions I mention. I suggest stepping though it line at a time and adjusting it for Fedora as you go. I'd be happy to help, it'd be something I'd use in the future too. 

With my ARM installations on Fedora, there's a few things I do a bit on top of the scripted install. I use MakeMKV with ARM, which has been a fantastic piece of software for ripping my discs, I use NVENC hardware acceleration with HandBrakeCLI, and recently I use FileBot to move the finished rips into my Plex libraries. (See [Call external program after completion](https://github.com/automatic-ripping-machine/automatic-ripping-machine/discussions/762))

**Reply by shitwolfymakes** [reply]
We recently decided to drop support for anything other than the docker distribution. 

That being said, a Fedora version of the `docker-setup.sh` script would be incredibly helpful

**Reply by PriamX** [reply]
Oh? What specifically are you looking for in a Fedora-based docker-setup script? As I mention somewhere in this thread I run ARM both directly on Fedora (no docker), and run ARM as a docker container on a Fedora-based host. I can tell you the container you have now works great as-is under Fedora. Really, the only trick was setting up the Nvidia container runtime for HW acceleration, which, IMHO, would be a setup on the side. This wasn't exclusive for ARM, I also use the runtime for my Plex and Tdarr containers. I'm happy to share any tricks, or help with a Fedora docker-setup.sh, but from what I've seen y'all have done a good job with the docker container and it works quite well on a Fedora-based docker host.

**Reply by shitwolfymakes** [reply]
Basically all I need is a copy of the script that is updated to use the correct package manager and is confirmed working. As I understand it Fedora doesn't use aptitude

**Reply by PriamX** [reply]
Yes, that's correct, RedHat variants like Fedora use `dnf`. The script I posted below should do the trick. However, it does need to be tested on a fresh Fedora install, I'll spool one up in a VM and clean it up.

**Reply by aclater** [reply]
How will the docker only variant handle environments where multiple CDroms are available (sometimes) ie - if I'm ripping one CD, I may only power up two of my external CD/DVD's, but if Ripping six - i'll start up all six of them?

**Reply by aclater** [reply]
Also - for fedora, lets use podman rather than introducing an additional repo dependency.

**Reply by PriamX** [reply]
> How will the docker only variant handle environments where multiple CDroms are available (sometimes) ie - if I'm ripping one CD, I may only power up two of my external CD/DVD's, but if Ripping six - i'll start up all six of them?

I've only ever used ARM with MakeMKV, in that case the MakeMKV software handles the optical drives. The arm docker runs in privileged mode, In the docker compose file (or run command), all the optical drives (/dev/srX) that you'd like the container to see would need to be exposed from the host to the container. If those devices are exposed through, then MakeMKV will see them.

**Reply by PriamX** [reply]
> Also - for fedora, lets use podman rather than introducing an additional repo dependency.

I just copied the concepts in the existing Ubuntu script which introduces a repo dependency for the makemkv software. I followed the same concept. However, podman could be used for both Fedora and Ubuntu.

**Reply by aclater** [reply]
> > How will the docker only variant handle environments where multiple CDroms are available (sometimes) ie - if I'm ripping one CD, I may only power up two of my external CD/DVD's, but if Ripping six - i'll start up all six of them?
> 
> I've only ever used ARM with MakeMKV, in that case the MakeMKV software handles the optical drives. The arm docker runs in privileged mode, In the docker compose file (or run command), all the optical drives (/dev/srX) that you'd like the container to see would need to be exposed from the host to the container. If those devices are exposed through, then MakeMKV will see them.

That's unfortunate. Has any work been done to eliminate the need for privileged mode? That would be a no-go for many systems.

**Reply by 1337-server** [reply]
> > > How will the docker only variant handle environments where multiple CDroms are available (sometimes) ie - if I'm ripping one CD, I may only power up two of my external CD/DVD's, but if Ripping six - i'll start up all six of them?
> > 
> > 
> > I've only ever used ARM with MakeMKV, in that case the MakeMKV software handles the optical drives. The arm docker runs in privileged mode, In the docker compose file (or run command), all the optical drives (/dev/srX) that you'd like the container to see would need to be exposed from the host to the container. If those devices are exposed through, then MakeMKV will see them.
> 
> That's unfortunate. Has any work been done to eliminate the need for privileged mode? That would be a no-go for many systems.

Already in progress. It needs more testing, but it can be done successfully.
I already have a PR open for this, I just haven't had enough time to test it fully.

**Reply by shitwolfymakes** [reply]
> How will the docker only variant handle environments where multiple CDroms are available (sometimes) ie - if I'm ripping one CD, I may only power up two of my external CD/DVD's, but if Ripping six - i'll start up all six of them? 

@aclater As long as you map all of the drives you want the container to have in the `docker run`, the container should be able to use them whenever they're plugged in

**Comment by PriamX** [comment]
Here's an updated and semi-tested Fedora 37 install script. Based on the current `ubuntu-20.04-install.sh`, some of @aclater 's script, and a few resolutions of my own. I have not tested this on a fresh Fedora, I tested it on a previously installed Fedora system I already had sitting around; milage may vary, let me know what issues you find.
```
#!/bin/bash
#
# fedora-37.01-install.sh
#
# Used the ubuntu-20.04-install.sh script as a template to start with.

set -eo pipefail

RED='\033[1;31m'
NC='\033[0m' # No Color
PORT=8080

function usage() {
    echo -e "\nUsage: $0 [OPTIONS]"
    echo -e " -p <port>\tSpecify the port arm will serve on. \n\t\tDefault is \"$PORT\""
    echo -e " -r\t\tOptionally prompt to set the optical disc\n\t\tdrive's region code."
}

port_flag=
while getopts 'p:r' OPTION
do
    case $OPTION in
    p)    port_flag=1
          # test if port is valid (DOES NOT WORK WITH `set -u` DECLARED)
          if ! [[ $OPTARG -gt 1023 && $OPTARG -le 65535 ]]; then
              echo -e "\nERROR: ${OPTARG} is not a valid port"
              usage
              exit 2
          fi
          PORT=$OPTARG
          ;;
    r)    region_set=1
          ;;
    ?)    usage
          exit 1
          ;;
    esac
done

function install_os_tools() {
    echo -e "${RED}Installing required OS tools${NC}"
    sudo dnf -yq upgrade
    sudo dnf install -y alsa-lib pipewire-alsa # this will install sound drivers if not already
    sudo dnf install -y lsscsi net-tools rsyslog
    sudo dnf install -y avahi && sudo systemctl restart avahi-daemon
    sudo dnf install -y git
}

function install_region_set() {
    # Optionally set the DVD region code
    # https://en.wikipedia.org/wiki/DVD_region_code
    if [ $region_set ]; then
        sudo dnf install -y regionset
        sudo regionset /dev/sr0
    fi
}

function add_arm_user() {
    echo -e "${RED}Adding arm user${NC}"
    # create arm group if it doesn't already exist
    if ! [[ "$(getent group arm)" ]]; then
        sudo groupadd arm
    else
        echo -e "${RED}arm group already exists, skipping...${NC}"
    fi

    # create arm user if it doesn't already exist
    if ! id arm >/dev/null 2>&1; then
        sudo useradd -m arm -g arm
        sudo passwd arm
    else
        echo -e "${RED}arm user already exists, skipping creation...${NC}"
    fi
    sudo usermod -aG cdrom,video arm
}

function install_arm_requirements() {
    echo -e "${RED}Installing ARM requirments${NC}"

    echo -e "${RED}Installing rpmfusion-free repository${NC}"
    sudo dnf -y install https://mirrors.rpmfusion.org/free/fedora/rpmfusion-free-release-"$(rpm -E %fedora)".noarch.rpm
    sudo dnf install rpmfusion-free-release-tainted # Needed for libdvdcss

    echo -e "${RED}Installing rpmfusion-nonfree repository${NC}"
    sudo dnf -y install https://mirrors.rpmfusion.org/nonfree/fedora/rpmfusion-nonfree-release-"$(rpm -E %fedora)".noarch.rpm

    sudo dnf update -y

    echo -e "${RED}Installing Fedora's Development Tools group${NC}"
    sudo dnf -y groupinstall "Development Tools"

    sudo dnf install -y curl libcurl libcurl-devel
    sudo dnf install -y openssl openssl-devel
    sudo dnf install -y systemd-udev libudev-devel

    sudo dnf install -y \
        python3 \
        python3-devel \
        python3-pip \
        python3-wheel \
        python-psutil \
        python3-pyudev \
        python3-netifaces \
        python3-testresources \
        python3-eyed3

    sudo dnf install -y qt5-qtbase-devel

    sudo dnf install -y abcde
    sudo dnf install -y AtomicParsley
    sudo dnf install -y cdparanoia cdparanoia-libs cdparanoia-static cdparanoia-devel
    sudo dnf install -y eject
    sudo dnf install -y ffmpeg ffmpeg-libs ffmpeg-devel
    sudo dnf install -y flac flac-libs flac-devel
    sudo dnf install -y glyr glyr-libs
    sudo dnf install -y java-openjdk-headless
    sudo dnf install -y libavcodec-free
    sudo dnf install -y expat expat-devel

    sudo dnf install -y HandBrake
    sudo dnf install -y ImageMagick ImageMagick-libs ImageMagick-devel
    sudo dnf install -y at
    sudo dnf install -y libdvdread libdvdread-devel
    sudo dnf install -y libdvdnav libdvdnav-devel
    sudo dnf install -y libdvdcss libdvdcss-devel
    sudo dnf install -y lsdvd
    sudo dnf install -y zlib zlib-devel

    # We'll use a 3rd party repo to install MakeMKV, this repo
    # may or may not be desired long-term, so we'll disable
    # afterwords to be on the safe side.
    sudo dnf config-manager --add-repo=https://negativo17.org/repos/fedora-multimedia.repo
    sudo dnf install -y makemkv
    sudo dnf config-manager --set-disabled fedora-multimedia

    # create folders required to run the ARM service
    sudo -u arm mkdir -p /home/arm/logs
}

function remove_existing_arm() {
    ##### Check if the ArmUI service exists in any state and remove it
    if sudo systemctl list-unit-files --type service | grep -F armui.service; then
        echo -e "${RED}Previous installation of ARM service found. Removing...${NC}"
        service=armui.service
        sudo systemctl stop $service && sudo systemctl disable $service
        sudo find /etc/systemd/system/$service -delete
        sudo systemctl daemon-reload && sudo systemctl reset-failed
    fi
}

function clone_arm() {
    cd /opt
    if [ -d arm ]; then
        echo -e "${RED}Existing ARM installation found, removing...${NC}"
        sudo rm -rf arm
    fi

    echo -e "${RED}Installing ARM from github...${NC}"
    git clone --recurse-submodules https://github.com/automatic-ripping-machine/automatic-ripping-machine --branch "main" arm

    cd arm
    git submodule update --init --recursive
    git submodule update --recursive --remote
    cd ..

    sudo chown -R arm:arm /opt/arm
    sudo find /opt/arm/scripts/ -type f -iname "*.sh" -exec chmod +x {} \;
}

function install_arm_env() {
    clone_arm
    echo -e "${RED}Installing ARM python dependencies...${NC}"
    cd /opt/arm
    pip3 install wheel
    pip3 install -r requirements.txt
}

function setup_config_files() {
    ##### Setup ARM-specific config files if not found
    sudo mkdir -p /etc/arm/config
    CONFS="arm.yaml apprise.yaml"
    for conf in $CONFS; do
        thisConf="/etc/arm/config/${conf}"
        if [[ ! -f "${thisConf}" ]] ; then
            echo "creating config file ${thisConf}"
            # Don't overwrite with defaults during reinstall
            cp --no-clobber "/opt/arm/setup/${conf}" "${thisConf}"
        fi
    done
    chown -R arm:arm /etc/arm/

    # abcde.conf is expected in /etc by the abcde installation
    cp --no-clobber "/opt/arm/setup/.abcde.conf" "/etc/.abcde.conf"
    chown arm:arm "/etc/.abcde.conf"
    # link to the new install location so runui.py doesn't break
    sudo -u arm ln -sf /etc/.abcde.conf /etc/arm/config/abcde.conf

    if [[ $port_flag ]]; then
        echo -e "${RED}Non-default port specified, updating arm config...${NC}"
        # replace the default 8080 port with the specified port
        sed -E s"/(^WEBSERVER_PORT:) 8080/\1 ${PORT}/" -i /etc/arm/config/arm.yaml
    else
        # reset the port number in the config since it's no longer being
        # overwritten which each run of this installer
        sed -E s"/(^WEBSERVER_PORT:) [0-9]+/\1 8080/" -i /etc/arm/config/arm.yaml
    fi

    # Grab the current MakeMKV key
    /opt/arm/scripts/update_key.sh
}

function setup_autoplay() {
    ##### Add new line to fstab, needed for the autoplay to work.
    echo -e "${RED}Adding fstab entry and creating mount points${NC}"
    for dev in /dev/sr?; do
        if grep -q "${dev}  /mnt${dev}  udf,iso9660  users,noauto,exec,utf8  0  0" /etc/fstab; then
            echo -e "${RED}fstab entry for ${dev} already exists. Skipping...${NC}"
        else
            echo -e "\n${dev}  /mnt${dev}  udf,iso9660  users,noauto,exec,utf8  0  0 \n" | sudo tee -a /etc/fstab
        fi
        sudo mkdir -p "/mnt$dev"
        sudo chown arm:arm "/mnt$dev"
    done

    # Put a rule in for udev
    sudo cp /opt/arm/setup/51-automedia.rules /etc/udev/rules.d/
    udevadm control --reload-rules && udevadm trigger
}

function setup_syslog_rule() {
    ##### Add syslog rule to route all ARM system logs to /var/log/arm.log
    if [ -f /etc/rsyslog.d/30-arm.conf ]; then
        echo -e "${RED}ARM syslog rule found. Overwriting...${NC}"
        sudo rm /etc/rsyslog.d/30-arm.conf
    fi
    sudo cp /opt/arm/setup/30-arm.conf /etc/rsyslog.d/30-arm.conf
    sudo chown arm:arm /etc/rsyslog.d/30-arm.conf
    sudo systemctl restart rsyslog
}

function install_armui_service() {
    ##### Install the ArmUI service
    echo -e "${RED}Installing ARM service${NC}"
    sudo mkdir -p /etc/systemd/system
    sudo cp /opt/arm/setup/armui.service /etc/systemd/system/armui.service
    sudo chmod 644 /etc/systemd/system/armui.service

    # reload the daemon and then start service
    sudo systemctl daemon-reload
    sudo systemctl start armui.service
    sudo systemctl enable armui.service
    sudo sysctl -p
}

function launch_setup() {
    echo -e "${RED}Launching ArmUI first-time setup${NC}"
    echo "Giving ArmUI a moment to start, standby..."
    sleep 30
    site_addr=$(sudo netstat -tlpn | awk '{ print $4 }' | grep ".*:${PORT}") || true
    if [[ -z "$site_addr" ]]; then
        echo -e "${RED}ERROR: ArmUI site is not running. Run \"sudo systemctl status armui\" to find out why${NC}"
    else
        echo -e "${RED}ArmUI site is running on http://$site_addr. Launching setup...${NC}"
        sudo -u arm nohup xdg-open "http://$site_addr/setup" > /dev/null 2>&1 &
    fi
}

function create_folders() {
    echo -e "${RED}Creating ARM folders${NC}"
    arm_mkdir "/home/arm/media/transcode"
    arm_mkdir "/home/arm/media/completed"
    arm_mkdir "/home/arm/media/raw"
}

function arm_mkdir() {
    echo -e "Creating $1"
    su - arm -c "mkdir -p -v $1"
}

# start here
install_os_tools
install_region_set
add_arm_user
install_arm_requirements
remove_existing_arm
install_arm_env
setup_config_files
setup_autoplay
setup_syslog_rule
install_armui_service
create_folders
launch_setup

#advise to reboot
echo
echo -e "${RED}We recommend rebooting your system at this time.${NC}"
echo

exit 0
```

**Reply by HaroldRobi** [reply]
that looks interesting.  great work

**Comment by petruut** [comment]
great work - will give this a go (as a total noob)

**Comment by aclater** [comment]
would anyone be opposed to using dnf -yq, also DNF is painfully slow - could we consolidate some of those dnf calls to a single one to speed things up a little bit? I'm currently testing on a fresh build of f38

re: docker/podman, I'd recommend deploying this as a [quadlet](https://github.com/containers/quadlet), running as the arm user.

a notional quadlet would look something like (based on https://github.com/automatic-ripping-machine/automatic-ripping-machine/blob/main/scripts/docker/start_arm_container.sh) & be started by systemd from /home/arm/.config/containers/systemd/arm-rippers.container :

```
# arm-rippers.container
[Container]
Image=IMAGE_NAME
AddDevice=/dev/sr0:/dev/sr0
AddDevice=/dev/sr1:/dev/sr1
AddDevice=/dev/sr2:/dev/sr2
AddDevice=/dev/sr3:/dev/sr3
ContainerName=arm-rippers
Environment="ARM_UID=<id -u arm>" "ARM_GID=<id -g arm>"
PublishPort=8080:8080
Volume=<path_to_arm_user_home_folder>:/home/arm
Volume=<path_to_music_folder>:/home/arm/music
Volume=<path_to_logs_folder>:/home/arm/logs
Volume=<path_to_media_folder>:/home/arm/media
Volume=<path_to_config_folder>:/etc/arm/config
PodmanArgs=--cpuset-cpus 2,3,4,5,6,7... --privileged

[Service]
Restart=always
```

**Comment by ewoks** [comment]


**Reply by mihawk90** [reply]
How about just using the Docker container since Fedora comes with Podman preinstalled?

Alternatively you could just scroll up.

That aside, if you want to install on an unsupported distro, it's on you to figure it out. It's not that hard to follow the instructions and replace Ubuntu-isms with Fedora-isms.

**Reply by ewoks** [reply]
@mihawk90 I wonder why such a unhelpful answer despite "it's not that hard". It does't contribute discussion or future questions, at all.

For the records:
- unclarities come from the fact that project focus is running arm with docker, but the documentation for it refers OS specific script
- just `docker run` of the latest image it's not enough for the basic ARM functionality 
- above script modifies host system quite a bit and it's few years old
- documentation doesn't mention in any section about many details which ubuntu script (or script above) does, although these are necessary

**Comment by HaroldRobi** [comment]
Yup you can use docker. The install documentation is on there GitHub site.
I just reinstalled my ARM using docker. Much better than the install script

Regards

-------------
"We've all seen that bumper sticker:    If God rode a motorcycle, it would
be a Harley.  I thought to myself...... now I know why God
hasn't made His second coming.....  He's still stuck in a cornfield out in
Nebraska or somewhere waiting for parts!"

**Reply by ewoks** [reply]
Thanks @HaroldRobi ... 
could you please confirm if `docker run ...` works for you on the clean machine or just on the one where some script was previously used?

**Comment by HaroldRobi** [comment]
Docker runs in its own container so a slightly used OS should not be a
problem.

Mine was using the current Ubuntu LTS release.  Installed manually using
the docker script. It took me about an hour.

Just to let you know that the team supporting ARM does not support manually
installing ARM. So the script documentation is out dated.  Best to use the
docker version.

**Reply by ewoks** [reply]
I'm interested in docker usage, but it seems the scripts are necessary for it to make it work and there I run into the issues...  :(

**Reply by ewoks** [reply]
e.g. I made the ARM start, device is visible in UI, but eject button from WebUI doesn't work although device is mounted to `/mnt/dev/sr0` and owned by `arm` user
```
eject: device name is `/dev/sr0'
eject: /dev/sr0: not mounted
eject: /dev/sr0: is whole-disk device
eject: /dev/sr0: closing tray
eject: cannot open /dev/sr0: Permission denied
```
After opening manually and putting disk inside, I see these two lines in log, but nothing happens after that
```
Mar 16 10:44:00 3f6eb7903a54 ARM: [ARM] Entering docker wrapper
Mar 16 10:44:00 3f6eb7903a54 ARM: [ARM] Starting ARM for DVD on sr0
```

**Comment by HaroldRobi** [comment]
Can you try from the command line of Fedora. Eject /dev/sr0

**Reply by ewoks** [reply]
already tried, that works without issues
